lib/rgeo/cartesian/calculations.rb

```diff
@@ -69,47 +69,236 @@ def contains_point?(p)
+                int_pt.distance(pt) < int_pt.distance(closest) ? pt : closest
+              end
+            end
+          end
         end
       end
 
       def length
         Math.sqrt(@lensq)
       end
     end
+
+    # Implements a Sweepline intersector to find all intersections
+    # in a group of segments. The idea is to use a horizontal line starting
+    # at y = +Infinity that sweeps down to y = -Infinity and every time it hits
+    # a new line, it will check if it intersects with any of the segments
+    # the line currently intersects at that y value.
+    # This is a more simplistic implementation that uses an array to hold
+    # observed segments instead of a sorted BST, so performance may be significantly
+    # worse in the case of lots of segments overlapping in y-ranges.
+    class SweeplineIntersector
+      Event = Struct.new(:point, :segment, :is_start)
+      Intersection = Struct.new(:point, :s1, :s2)
+
+      def initialize(segments)
+        @segments = segments
+        @events = []
+        @intersections = []
+        @proper_intersections = []
+      end
+      attr_reader :segments
+
+      # Returns the "proper" intersections from the list of segments.
+      #
+      # This will only return intersections that are not the start/end or 
+      # end/start of the 2 segments. This could be useful for finding intersections
+      # in a ring for example, because knowing that segments are connected in a linestring
+      # is not always helpful, but those are reported by default.
+      #
+      # Note: This is not the true definition of a proper intersection. A
+      # truly proper intersection does not include colinear intersections and
+      # the intersection must lie in the interior of both segments.
+      #
+      # @return [Array<RGeo::Cartesian::SweeplineIntersector::Intersection>]
+      def proper_intersections
+        return @proper_intersections if @proper_intersections.size.positive?
+
+        intersections.each do |intersection|
+          s1 = intersection.s1
+          s2 = intersection.s2
+          pt = intersection.point
+
+          unless (pt == s1.s && pt == s2.e) || (pt == s1.e && pt == s2.s)
+            @proper_intersections << intersection
+          end
+        end
+        @proper_intersections
+      end
+
+      # Computes the intersections of the input segments.
+      #
+      # Creates an event queue from the +events+ and adds segments to the
+      # +observed_segments+ array while their ending event has not been popped
+      # from the queue.
+      #
+      # Compares the new segment from the +is_start+ event to each observed segment
+      # then adds it to +observed_segments+. Records any intersections in to the
+      # returned array.
+      #
+      # @return [Array<RGeo::Cartesian::SweeplineIntersector::Intersection>]
+      def intersections
+        return @intersections if @intersections.size.positive?
```

Comment:
rather than initializing and checking for the size, you could check for definition. It is a bit faster, and IMHO clearer on intention:

```suggestion
        return @intersections if defined?(@intersections)
```